File: 2021/python/day12.py
from collections import defaultdict


class Graph:
    def __init__(self):
        self.graph = defaultdict(list)
        self.paths = []
        self.current_path = []
        self.visited = defaultdict(int)
        self.abort = lambda v, n: v[n] == 1

    def add_edge(self, a, b):
        self.graph[a].append(b)

    # https://www.baeldung.com/cs/simple-paths-between-two-vertices
    def dfs(self, start, end):
        if self.abort(self.visited, start):
            return
        if start.islower():
            self.visited[start] += 1
        self.current_path.append(start)
        if start == end:
            self.paths.append(self.current_path.copy())
            self.visited[start] -= 1
            self.current_path.pop()
            return
        for neighbor in self.graph[start]:
            self.dfs(neighbor, end)
        self.current_path.pop()
        self.visited[start] -= 1


def load_file():
    with open('../inputs/input12') as file:
        return file.read()


def create_graph():
    data = load_file()
    graph = Graph()
    for line in data.splitlines():
        a, b = line.split('-')
        graph.add_edge(a, b)
        graph.add_edge(b, a)
    return graph


def part1():
    graph = create_graph()
    graph.dfs('start', 'end')
    return len(graph.paths)


def part2():
    def abort(visited, node):
        visited_twice = sum([1 for v in visited.values() if v >= 2])
        if visited_twice:
            if visited[node] >= 1:
                return True
        elif node in ['start', 'end'] and visited[node] >= 1:
            return True
        return False

    graph = create_graph()
    graph.abort = abort
    graph.dfs('start', 'end')
    return len(graph.paths)


print(part1())  # 4549
print(part2())  # 120535

File: 2021/rust/src/day12.rs
use std::collections::HashMap;

struct Graph<'a> {
    graph: HashMap<&'a str, Vec<&'a str>>,
    visited: HashMap<&'a str, isize>,
    current_path: Vec<&'a str>,
    paths: Vec<Vec<&'a str>>,
}

impl<'a> Graph<'a> {
    fn new() -> Self {
        Self {
            graph: HashMap::<&str, Vec<&str>>::new(),
            visited: HashMap::<&str, isize>::new(),
            current_path: vec![],
            paths: vec![],
        }
    }

    fn add_edge(&mut self, a: &'a str, b: &'a str) {
        self.graph.entry(a).or_insert(vec![]);
        self.graph.get_mut(a).unwrap().push(b);
        self.graph.entry(b).or_insert(vec![]);
        self.graph.get_mut(b).unwrap().push(a);
    }

    fn dfs(
        &mut self,
        start: &'a str,
        end: &'a str,
        abort: &dyn Fn(&HashMap<&str, isize>, &str) -> bool,
    ) {
        if abort(&self.visited, start) {
            return;
        }

        if start.chars().all(|c| c.is_lowercase()) {
            *self.visited.entry(start).or_insert(0) += 1;
        }
        self.current_path.push(start);
        if start == end {
            self.paths.push(self.current_path.clone());
            *self.visited.entry(start).or_insert(0) -= 1;
            self.current_path.pop();
            return;
        }
        for neighbor in &self.graph[start].clone() {
            self.dfs(neighbor, end, &abort);
        }
        self.current_path.pop();
        *self.visited.entry(start).or_insert(0) -= 1;
    }

    fn count_paths(&self) -> usize {
        self.paths.len()
    }
}

fn create_graph(lines: &Vec<String>) -> Graph {
    let mut graph = Graph::new();
    for line in lines {
        let tokens = line.split('-').collect::<Vec<&str>>();
        graph.add_edge(tokens[0], tokens[1]);
    }
    graph
}

pub fn part1(lines: &Vec<String>) -> usize {
    let abort = |v: &HashMap<&str, isize>, n: &str| v.contains_key(n) && v[n] == 1;
    let mut graph = create_graph(lines);
    graph.dfs("start", "end", &abort);
    graph.count_paths()
}

pub fn part2(lines: &Vec<String>) -> usize {
    let abort = |v: &HashMap<&str, isize>, n: &str| {
        let visited_twice = v.values().any(|v| *v >= 2);
        if visited_twice && v.contains_key(n) && v[n] >= 1 {
            return true;
        } else if (n == "start" || n == "end") && v.contains_key(n) && v[n] >= 1 {
            return true;
        }
        false
    };
    let mut graph = create_graph(lines);
    graph.dfs("start", "end", &abort);
    graph.count_paths()
}
